### PR TITLE
feat(codegen): add codegen skeleton package

### PR DESCRIPTION
**Pull request by sgulseth**
### Description

This PR introduces a skeleton for the new codegen package, it should just introduce the infrastructure and tooling for the package.
Features are introduced in the subsequent stacked PRs.

I will wait to merge the PRs until all prs are approved and ready ✅ 

### What to review

* Package setup and general configuration

### Testing

N/A

### Notes for release

N/A - not to include in release notes

### CODE REVIEW

**Comment by sgulseth** [comment]
* **#5982** <a href="https://app.graphite.dev/github/pr/sanity-io/sanity/5982?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#5981** <a href="https://app.graphite.dev/github/pr/sanity-io/sanity/5981?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#5980** <a href="https://app.graphite.dev/github/pr/sanity-io/sanity/5980?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#5979** <a href="https://app.graphite.dev/github/pr/sanity-io/sanity/5979?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `next`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @sgulseth and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/@sanity/codegen/package.json [hunk 1]
```diff
@@ -0,0 +1,64 @@
+{
+  "name": "@sanity/codegen",
+  "version": "3.30.0",
```
Comment (bjoerge) [inline packages/@sanity/codegen/package.json:3]: this should be updated to `3.33.0`

File: packages/@sanity/codegen/package.json
Comment (bjoerge) [inline packages/@sanity/codegen/package.json:1]: You probably want to explicity set publishConfig to make the package public once it's released (scoped packages are restricted by default)
See example here: https://github.com/sanity-io/sanity/commit/c4063d30fe102d29de9096f0cc670e60a7fea0d6#diff-f5810e1e856b0a198e1b018aa1be873b99ca708a022833465c9fb553fb8df0ecR40-R42

File: packages/@sanity/codegen/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  "include": ["./src/**/*"],
+  "compilerOptions": {
+    "composite": true,
+    "lib": ["ES2022.Error"],
+    "rootDir": ".",
+    "outDir": "./lib/dts",
+    "declarationDir": "./lib/dts"
+  },
+  "references": [
+    {"path": "../../@sanity/types/tsconfig.lib.json"},
```
Comment (bjoerge) [inline packages/@sanity/codegen/tsconfig.json:12]:
```suggestion
    {"path": "../types/tsconfig.lib.json"},
```

File: packages/@sanity/codegen/tsconfig.lib.json [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  "include": ["./src/**/*"],
+  "compilerOptions": {
+    "composite": true,
+    "lib": ["ES2022.Error"],
+    "rootDir": ".",
+    "outDir": "./lib/dts",
+    "declarationDir": "./lib/dts"
+  },
+  "references": [
+    {"path": "../../@sanity/types/tsconfig.lib.json"},
```
Comment (bjoerge) [inline packages/@sanity/codegen/tsconfig.lib.json:12]:
```suggestion
    {"path": "../types/tsconfig.lib.json"},
```

**Review by ricokahler — APPROVED** [review]
Looks good!

**Review by bjoerge — APPROVED** [review]

**Pull request merged**